### PR TITLE
feat(eventing): add io-engine events

### PR DESCRIPTION
**Pull request by datacore-vvarakantham**
Added io-engine start, stop, shutdown events.
Reactor freeze, unfreeze events.
Nexus shutdown event.

### CODE REVIEW

File: io-engine/src/core/env.rs [hunk 1]
```diff
@@ -466,6 +468,9 @@ async fn do_shutdown(arg: *mut c_void) {
         spdk_rpc_finish();
         spdk_subsystem_fini(Some(reactors_stop), arg);
     }
+    MayastorEnvironment::global_or_default()
```
Comment (tiagolobocastro) [inline io-engine/src/core/env.rs]: Shouldn't this be generated at the start of shutdown?

File: io-engine/src/core/env.rs [hunk 2]
```diff
@@ -477,6 +482,9 @@ pub fn mayastor_env_stop(rc: i32) {
             r.send_future(async move {
                 do_shutdown(rc as *const i32 as *mut c_void).await;
             });
+            MayastorEnvironment::global_or_default()
+                .event(EventAction::Stop)
+                .generate();
```
Comment (tiagolobocastro) [inline io-engine/src/core/env.rs]: Shouldn't this be generated right before we stop only?
And would it be worth it to include how long it took to stop?

Comment (datacore-vvarakantham) [reply io-engine/src/core/env.rs]: Yes. Should I add the time taken to stop field in nanoseconds time?

Comment (tiagolobocastro) [reply io-engine/src/core/env.rs]: Can just use `Duration` I guess?

Comment (datacore-vvarakantham) [reply io-engine/src/core/env.rs]: Yes, I can use the Duration.

Comment (datacore-vvarakantham) [reply io-engine/src/core/env.rs]: Fixed. Added the time taken to event specific details.

File: io-engine/src/core/env.rs (hunk 1 above)
Comment (datacore-vvarakantham) [reply io-engine/src/core/env.rs]: Added the event at the start.

File: io-engine/src/eventing/mod.rs [hunk 3]
```diff
@@ -1,12 +1,13 @@
...
+pub(crate) mod io_engine_events;
 mod nexus_child_events;
 pub(crate) mod nexus_events;
 mod pool_events;
 mod replica_events;
 use events_api::event::{EventAction, EventMessage, EventMeta};
 
 /// Event trait definition for creating events.
-pub(crate) trait Event {
+pub trait Event {
```
Comment (tiagolobocastro) [inline io-engine/src/eventing/mod.rs:10]: Does it need to be pub?

**Review by tiagolobocastro — APPROVED** [review]

File: io-engine/src/bdev/nexus/mod.rs [hunk 4]
```diff
@@ -173,15 +174,19 @@ pub async fn shutdown_nexuses() {
...
-                name = nexus.name,
-                error = error.verbose(),
-                "Failed to destroy nexus"
-            );
+        match nexus.as_mut().destroy_ext(true).await {
+            Ok(_) => {
+                nexus.event(EventAction::Shutdown).generate();
+            }
+            Err(error) => {
+                error!(
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nexus/mod.rs:185]: Wonder if maybe shutdown can include a result?

File: io-engine/src/eventing/mod.rs (hunk 3 above)
Comment (datacore-vvarakantham) [reply io-engine/src/eventing/mod.rs:10]: It does not need to be pub here. pub(crate) works as all the events are in the same crate.

File: io-engine/src/bdev/nexus/mod.rs (hunk 4 above)
Comment (datacore-vvarakantham) [reply io-engine/src/bdev/nexus/mod.rs:185]: What would the result contain here? If the shutdown is failed or successful?

Comment (tiagolobocastro) [reply io-engine/src/bdev/nexus/mod.rs:185]: Yes

Comment (datacore-vvarakantham) [reply io-engine/src/bdev/nexus/mod.rs:185]: Added the error to event details.

**Review by sinhaashish — APPROVED** [review]

**Comment by datacore-vvarakantham** [comment]
bors merge

**Pull request merged**